aten/src/ATen/native/cuda/TensorFactories.cu

```diff
@@ -26,4 +36,54 @@ Tensor& eye_out_cuda(Tensor& result, int64_t n, int64_t m) {
   return result;
 }
 
+Tensor& randperm_out_cuda(Tensor& result, int64_t n, Generator* generator) {
+  if (n < 0) {
+    std::ostringstream oss;
+    oss << "n must be non-negative, got " << n;
+    throw std::runtime_error(oss.str());
+  }
+
+  auto result_tmp = result;
+  if (result.type().scalarType() == at::ScalarType::Half) {
+    // Make sure n is within range of Half
+    assert(Scalar(n).toHalf());
+    result_tmp = CUDA(kFloat).tensor();
+  }
```

Comment:
It would also be much simpler to avoid wrapping everything in here in `result_tmp`, but do everything inside this branch, using a recursive call.